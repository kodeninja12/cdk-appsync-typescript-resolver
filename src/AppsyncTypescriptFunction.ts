import * as appsync from 'aws-cdk-lib/aws-appsync';
import type { IConstruct } from 'constructs';
import { bundle } from './bundler';

export interface AppsyncTypescriptFunctionProps extends appsync.AppsyncFunctionProps {
  /**
   * Path of typescript file that will be transpiled and bundled
   */
  readonly path: string;
  /**
   * Flag to enable or disable source maps in bundled code, defaulted to false
   */
  readonly sourceMap?: boolean;
  /**
   * A map of replacement strings in the bundled code. Example: { "ENV", "prod" }
    */
  readonly replaceStrings?: Record<string, string>;
}

/**
 * Transpile and bundle Typescript to AWS Appsync JS function
 */
export class AppsyncTypescriptFunction extends appsync.AppsyncFunction {
  constructor(
    scope: IConstruct,
    id: string,
    props: AppsyncTypescriptFunctionProps,
  ) {
    const { path, sourceMap, replaceStrings } = props;
    super(scope, id, {
      ...props,
      code: appsync.Code.fromInline(
        bundle({
          entryPoint: path,
          sourceMap: sourceMap ?? false,
          replaceStrings,
        }),
      ),
      runtime: appsync.FunctionRuntime.JS_1_0_0,
    });
  }
}
