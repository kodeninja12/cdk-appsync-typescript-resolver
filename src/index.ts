export * from './AppsyncTypescriptFunction';
export * from './TSExpressPipelineResolver';
