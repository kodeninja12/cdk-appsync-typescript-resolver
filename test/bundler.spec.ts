import * as path from 'path';
import { bundle } from '../src/bundler';

describe('bundler', () => {
  test.each([true, false])(
    'should bundle ts to js with sourcemap set to %s',
    (sourceMap) => {
      const result = bundle({
        entryPoint: path.join(__dirname, 'resources', 'getTask.ts'),
        sourceMap,
      });

      expect(result).toMatchSnapshot();
    },
  );

  test('should bundle ts to js with replace strings', () => {
    const result = bundle({
      entryPoint: path.join(
        __dirname,
        'resources',
        'bundlerReplaceTest.ts',
      ),
      sourceMap: false,
      replaceStrings: {
        __buildVersion: '1.0.0',
        __STAGE: 'test',
      },
    });

    expect(result).toStrictEqual(
      '// test/resources/bundlerReplaceTest.ts\n' +
                'var getBuildVersion = () => {\n' +
                '  return "test::test::1.0.0";\n' +
                '};\n' +
                'export {\n' +
                '  getBuildVersion\n' +
                '};\n',
    );
  });
});
