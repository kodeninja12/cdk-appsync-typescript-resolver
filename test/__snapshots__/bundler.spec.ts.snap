// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundler should bundle ts to js with sourcemap set to false 1`] = `
"// test/resources/getTask.ts
import { util as util2 } from "@aws-appsync/utils";

// test/resources/utils.ts
import { util } from "@aws-appsync/utils";
var generateAutoId = () => util.autoId();

// test/resources/getTask.ts
function request(ctx) {
  console.log("ctx.args.id", ctx.args.id);
  return {
    operation: "GetItem",
    key: {
      id: util2.dynamodb.toDynamoDB(ctx.args.id)
    }
  };
}
function response(ctx) {
  return { ...ctx.result, id: generateAutoId() };
}
export {
  request,
  response
};
"
`;

exports[`bundler should bundle ts to js with sourcemap set to true 1`] = `
"// test/resources/getTask.ts
import { util as util2 } from "@aws-appsync/utils";

// test/resources/utils.ts
import { util } from "@aws-appsync/utils";
var generateAutoId = () => util.autoId();

// test/resources/getTask.ts
function request(ctx) {
  console.log("ctx.args.id", ctx.args.id);
  return {
    operation: "GetItem",
    key: {
      id: util2.dynamodb.toDynamoDB(ctx.args.id)
    }
  };
}
function response(ctx) {
  return { ...ctx.result, id: generateAutoId() };
}
export {
  request,
  response
};
//# sourceMappingURL=data:application/json;base64,ewogICJ2ZXJzaW9uIjogMywKICAic291cmNlcyI6IFsidGVzdC9yZXNvdXJjZXMvZ2V0VGFzay50cyIsICJ0ZXN0L3Jlc291cmNlcy91dGlscy50cyJdLAogICJtYXBwaW5ncyI6ICI7QUFBQSxTQUEwQyxRQUFBQSxhQUFZOzs7QUNBdEQsU0FBUyxZQUFZO0FBQ2QsSUFBTSxpQkFBaUIsTUFBTSxLQUFLLE9BQU87OztBREd6QyxTQUFTLFFBQ2QsS0FDd0I7QUFDeEIsVUFBUSxJQUFJLGVBQWUsSUFBSSxLQUFLLEVBQUU7QUFFdEMsU0FBTztBQUFBLElBQ0wsV0FBVztBQUFBLElBQ1gsS0FBSztBQUFBLE1BQ0gsSUFBSUMsTUFBSyxTQUFTLFdBQVcsSUFBSSxLQUFLLEVBQUU7QUFBQSxJQUMxQztBQUFBLEVBQ0Y7QUFDRjtBQUVPLFNBQVMsU0FDZCxLQUNBO0FBQ0EsU0FBTyxFQUFFLEdBQUcsSUFBSSxRQUFRLElBQUksZUFBZSxFQUFFO0FBQy9DOyIsCiAgIm5hbWVzIjogWyJ1dGlsIiwgInV0aWwiXQp9Cg==
"
`;
